pkg/kubelet/pleg/generic.go

```diff
@@ -116,40 +158,190 @@ func (g *GenericPLEG) relist() {
+			cid := oldContainer.ID
+			oldState := convertState(oldContainer.State)
+			newState := getContainerState(pod, cid)
+			if newState != plegContainerNonExistent {
+				// We already processed the container.
+				continue
+			}
+			// Container no longer visible, generate an event.
+			e := generateEvent(pod.ID, cid.ID, oldState, plegContainerNonExistent)
+			updateEvents(eventsByPodID, e)
+		}
+	}
+	// Process all pods that are no longer visible.
+	for pid := range g.podRecords {
+		if pod := g.podRecords.getCurrent(pid); pod != nil {
+			continue
+		}
+		oldPod := g.podRecords.getOld(pid)
+		for _, oldContainer := range oldPod.Containers {
+			cid := oldContainer.ID
+			oldState := convertState(oldContainer.State)
+			e := generateEvent(oldPod.ID, cid.ID, oldState, plegContainerNonExistent)
+			updateEvents(eventsByPodID, e)
+		}
+	}
 
-	events := []*PodLifecycleEvent{}
-	containers := make(map[string]containerInfo, len(g.containers))
-	// Create a new containers map, compares container statuses, and generates
-	// correspoinding events.
-	for _, p := range pods {
-		for _, c := range p.Containers {
-			cid := c.ID.ID
-			// Get the of existing container info. Defaults to state unknown.
-			oldState := kubecontainer.ContainerStateUnknown
-			if info, ok := g.containers[cid]; ok {
-				oldState = info.state
+	if g.cacheEnabled() {
+		// If there are events associated with a pod, we should update the
+		// podCache.
+		for pid, events := range eventsByPodID {
+			pod := g.podRecords.getCurrent(pid)
+			if pod == nil {
+				// The pod is missing in the current relist. This means that
+				// the pod has no visible (active or inactive) containers.
+				g.removeFromCache(pid)
+				continue
+			}
+			// updateCache() will inspect the pod and update the cache. If an
+			// error occurs during the inspection, we want PLEG to retry again
+			// in the next relist. To achieve this, we do not update the
+			// associated podRecord of the pod, so that the change will be
+			// detect again in the next relist.
+			// TODO: If many pods changed during the same relist period,
+			// inspecting the pod and getting the PodStatus to update the cache
+			// serially may take a while. We should be aware of this and
+			// parallelize if needed.
+			if err := g.updateCache(pod); err != nil {
+				glog.Errorf("PLEG: Ignoring events for pod %s/%s: %v", pod.Name, pod.Namespace, err)
+				continue
+			}
+			// Update the internal storage and send out the events.
+			g.podRecords.update(pid)
+			for i := range events {
+				g.eventChannel <- events[i]
 			}
-			// Generate an event if required.
-			glog.V(7).Infof("GenericPLEG: %v/%v: %v -> %v", p.ID, cid, oldState, c.State)
-			if e := generateEvent(p.ID, cid, oldState, c.State); e != nil {
-				events = append(events, e)
+		}
+		// Update the cache timestamp.  This needs to happen *after*
+		// all pods have been properly updated in the cache.
+		g.cache.UpdateTime(timestamp)
+	} else {
```

Comment:
conversation is missed here:

>yujuhong added a note a day ago
I think it's more clear to keep the if-else blocks since return the blocks are quite big -- people may miss the return statement.
 @yifan-gu  

>yifan-gu added a note 20 hours ago
@yujuhong hmm, how about moving the else part to the if part:

if !g.cacheEnabled() {
    g.podRecords.updateAll()
    for ... {
    }
    return
}

// Cache enabled.
for pid, events := range eventsByPodID {
    ....
}
g.cache.UpdateTime(timestamp)